### DISCUSSION TITLE
Milvus example index.go unexpected performance result

### DISCUSSION THREAD

**Discussion by joshuakosasih** [milvus-io/milvus#16012] [Q&A and General discussion]
Hi I'm interested in using milvus as now the v2.0 supports hybrid search feature. I give it a try by [setting up a local standalone milvus](https://milvus.io/docs/v2.0.x/install_standalone-docker.md) and run a [simple example given in milvus-sdk-go](https://github.com/milvus-io/milvus-sdk-go/blob/master/examples/index/index.go). The example logic flow is simple: 

1. setup data in milvus
2. search (with hybrid search)
3. calculate the latency
4. add index (ivfflat)
5. search again (same with step 2)
6. calculate latency

**The result is surprising**: latency after adding index on average **are the same** with before adding index, if not worse. There's no improvement whatsoever after adding index.
```
2022/02/25 16:49:55 #1
2022/02/25 16:49:59 search without index time elapsed: 337.428047ms
2022/02/25 16:50:03 search with index time elapsed: 363.08467ms
2022/02/25 16:50:04 #2
2022/02/25 16:50:07 search without index time elapsed: 261.659237ms
2022/02/25 16:50:11 search with index time elapsed: 203.145849ms
2022/02/25 16:50:12 #3
2022/02/25 16:50:16 search without index time elapsed: 260.500598ms
2022/02/25 16:50:18 search with index time elapsed: 351.080185ms
2022/02/25 16:50:19 #4
2022/02/25 16:50:22 search without index time elapsed: 268.742843ms
2022/02/25 16:50:27 search with index time elapsed: 376.56818ms
2022/02/25 16:50:28 #5
2022/02/25 16:50:31 search without index time elapsed: 249.382221ms
2022/02/25 16:50:36 search with index time elapsed: 341.482783ms
```

I also tried to search normally (not hybrid) by putting empty string "" to the bool expression arg in search function, but the result is also the same

```
2022/02/25 16:48:04 #1
2022/02/25 16:48:07 search without index time elapsed: 272.571346ms
2022/02/25 16:48:12 search with index time elapsed: 362.661952ms
2022/02/25 16:48:13 #2
2022/02/25 16:48:16 search without index time elapsed: 273.962792ms
2022/02/25 16:48:21 search with index time elapsed: 357.490357ms
2022/02/25 16:48:22 #3
2022/02/25 16:48:26 search without index time elapsed: 336.229492ms
2022/02/25 16:48:30 search with index time elapsed: 398.66543ms
2022/02/25 16:48:31 #4
2022/02/25 16:48:34 search without index time elapsed: 219.972767ms
2022/02/25 16:48:39 search with index time elapsed: 351.866163ms
2022/02/25 16:48:40 #5
2022/02/25 16:48:43 search without index time elapsed: 209.774731ms
2022/02/25 16:48:48 search with index time elapsed: 358.223902ms
```

Is this expected?

**Comment by xiaofan-luan** [comment]
hi @joshuakosasih , how large is the dataset? if data is too small then index won't help.
It could be super helpful if you share your code of test

**Reply by joshuakosasih** [reply]
hi, i used the example provided here https://github.com/milvus-io/milvus-sdk-go/blob/master/examples/index/index.go
it uses this file https://github.com/milvus-io/milvus-sdk-go/blob/master/examples/films.csv, so seems about 8k of data

**Comment by yanliang567** [comment]
besides the data size, I also curious about the index parameters and the search parameters

**Reply by joshuakosasih** [reply]
the index parameter & search parameter also follows the example https://github.com/milvus-io/milvus-sdk-go/blob/master/examples/index/index.go i almost did no change on it

**Comment by yhmo** [comment]
It seems like a "strong guarantee" search. So far as I now, the Go SDK doesn't support specify consistence level in search(). So each search request will cost 200ms - 400ms.

**Reply by joshuakosasih** [reply]
hi, sorry what is this consistency level that you mean?

**Reply by yhmo** [reply]
Milvus 2.x is a distributed system, it use timestamp to sync operations(insert, delete, etc.) across multiple nodes. Milvus internally has a time interval(200ms by default) to sync operations batch by batch. Search/query requests will see different data view, the "consistence level" is to define which data view can be saw by a search request.

In python sdk, the search() has a parameter "consistency_level", if you input "Strong", the search request will wait about 200ms to ensure all operations are synced successfully:
```
results = collection.search(
	data=[[0.1, 0.2]], 
	anns_field="book_intro", 
	param=search_params, 
	limit=10, 
	expr=None,
	consistency_level="Strong"
)
```

In Java/CPP sdk, it also has a similar parameter "guarantee_timestamp" to specify the "consistence_level". The Go sdk not yet expose it, so each search request will wait about 200ms.
